src/clang_parser.cpp

```diff
@@ -286,55 +292,80 @@ void ClangParser::parse(ast::Program *program, BPFtrace &bpftrace)
     std::cerr << "Input (" << input.size() << "): " << input << std::endl;
   }
 
+  indirect_structs.clear();
+  unvisited_indirect_structs.clear();
+
   CXCursor cursor = clang_getTranslationUnitCursor(translation_unit);
 
-  clang_visitChildren(
-      cursor,
-      [](CXCursor c, CXCursor parent, CXClientData client_data)
-      {
+  bool iterate = true;
 
-        if (clang_getCursorKind(c) == CXCursor_MacroDefinition)
+  do {
+    clang_visitChildren(
+        cursor,
+        [](CXCursor c, CXCursor parent, CXClientData client_data)
         {
-          std::string macro_name;
-          std::string macro_value;
-          if (translateMacro(c, macro_name, macro_value)) {
-            auto &macros = static_cast<BPFtrace*>(client_data)->macros_;
-            macros[macro_name] = macro_value;
+          if (clang_getCursorKind(c) == CXCursor_MacroDefinition)
+          {
+            std::string macro_name;
+            std::string macro_value;
+            if (translateMacro(c, macro_name, macro_value)) {
+              auto &macros = static_cast<BPFtrace*>(client_data)->macros_;
+              macros[macro_name] = macro_value;
+            }
+            return CXChildVisit_Recurse;
           }
-          return CXChildVisit_Recurse;
-        }
-
-        if (clang_getCursorKind(parent) != CXCursor_StructDecl &&
-            clang_getCursorKind(parent) != CXCursor_UnionDecl)
-          return CXChildVisit_Recurse;
 
-        if (clang_getCursorKind(c) == CXCursor_FieldDecl)
-        {
-          auto &structs = static_cast<BPFtrace*>(client_data)->structs_;
-          auto struct_name = get_parent_struct_name(c);
-          auto ident = get_clang_string(clang_getCursorSpelling(c));
-          auto offset = clang_Cursor_getOffsetOfField(c) / 8;
-          auto type = clang_getCanonicalType(clang_getCursorType(c));
+          if (clang_getCursorKind(parent) != CXCursor_StructDecl &&
+              clang_getCursorKind(parent) != CXCursor_UnionDecl)
+            return CXChildVisit_Recurse;
 
           auto ptype = clang_getCanonicalType(clang_getCursorType(parent));
           auto ptypestr = get_clang_string(clang_getTypeSpelling(ptype));
           auto ptypesize = clang_Type_getSizeOf(ptype);
 
-          if(clang_Cursor_isAnonymous(parent))
-            offset = get_indirect_field_offset(c);
+          if (clang_getCursorKind(c) == CXCursor_StructDecl ||
+              clang_getCursorKind(c) == CXCursor_UnionDecl) {
+            auto struct_name = get_clang_string(clang_getTypeSpelling(clang_getCanonicalType(clang_getCursorType(c))));
+            indirect_structs[struct_name] = c;
+            unvisited_indirect_structs.insert(struct_name);
 
-          if (struct_name == "")
-            struct_name = ptypestr;
-          remove_struct_prefix(struct_name);
+            return CXChildVisit_Continue;
```

Comment:
If we returned CXChildVisit_Recurse here can we get rid of the looping logic and recording univisited indirect structs? Sorry, my build is still broken so I can't check myself (LLVM 8 wasn't the only problem on my computer apparently)

Reply:
No because at this point we don't know if this struct is just being defined (indirect struct) or if it will be used as the type for a field of the parent struct (direct struct). The cursor is pointing to the struct declaration, not in the field declaration.

Look at the struct below (the current clang cursor is what's inside `<` and `>`:

```
struct Foo {
  int field1;
  <struct {
    int field2;
  }> field3;
  char field4;
}
```

In this case, we want to be able to access `field2` as `Foo.field3.field2`. But we don't know if this is a direct or indirect struct, and we'll only know once we move our cursor to forward, in which case we'll be at:


```
struct Foo {
  int field1;
  struct {
    int field2;
  } <field3>;
  char field4;
}
```

Same thing goes to indirect structs, we'll only know how the access will work once we move the cursor.

Older versions of LLVM would differentiate those two access types with clang_Cursor_isAnonymous, but from what I can tell this was a bug which was fixed on LLVM 8. Keeping track of inner structs and deferring their parsing until we have enough information was the only way I could find to keep things working across LLVM versions.